### ISSUE TITLE
 /usr/lib64/libc.so.6: version `GLIBC_2.18' not found 

### ISSUE THREAD

**Issue by lasorda** [withered-magic/starpls#334]
e4b563cedbc9c84bb5672559b1baa5e244652c39 don't has such problem，I see you [build: update rules_rust to 0.53.0](https://github.com/withered-magic/starpls/commit/7967acff66d7d7b4d0c3dfad07c24e5984932418) and [build: migrate to bzlmod ](https://github.com/withered-magic/starpls/commit/baba01c4fa6642402ac7703f90e01012bdbf3fc7)，I don't know how to fix it

**Comment by withered-magic** [comment]
Can you provide more information on what OS you're running? Additionally, when exactly do you get this error? Do you get it trying to run `bazel build` or while using the pre-built binaries?

**Comment by lasorda** [comment]
> Can you provide more information on what OS you're running? Additionally, when exactly do you get this error? Do you get it trying to run `bazel build` or while using the pre-built binaries?
> 
> Additionally can you run `ldd --version` to see what version of glibc you currently have on your system?

```
➜ ldd --version
ldd (GNU libc) 2.17
Copyright (C) 2012 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```
I use centos 7，I download the pre-built binaries
```
➜ ./starpls-linux-amd64
./starpls-linux-amd64: /usr/lib64/libc.so.6: version `GLIBC_2.18' not found (required by ./starpls-linux-amd64)
./starpls-linux-amd64: /usr/lib64/libc.so.6: version `GLIBC_2.25' not found (required by ./starpls-linux-amd64)
./starpls-linux-amd64: /usr/lib64/libc.so.6: version `GLIBC_2.28' not found (required by ./starpls-linux-amd64)
```
Build steps
```
➜ git clone git@github.com:withered-magic/starpls.git
➜ cd starpls
➜ wget https://github.com/bazelbuild/bazel/releases/download/7.4.1/bazel-7.4.1-linux-x86_64
➜ chmod +x bazel-7.4.1-linux-x86_64
➜ ./bazel-7.4.1-linux-x86_64 run -c opt //editors/code:copy_starpls
Starting local Bazel server and connecting to it...
ERROR: /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/external/rules_rust~/crate_universe/private/module_extensions/cargo_bazel_bootstrap.bzl:66:17: Traceback (most recent call last):
        File "/data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/external/rules_rust~/crate_universe/extension.bzl", line 369, column 37, in _crate_impl
                _generate_hub_and_spokes(module_ctx, cargo_bazel, cfg, annotations, cargo_lockfile = cargo_lockfile, manifests = manifests)
        File "/data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/external/rules_rust~/crate_universe/extension.bzl", line 132, column 16, in _generate_hub_and_spokes
                cargo_bazel(splice_args)
        File "/data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/external/rules_rust~/crate_universe/private/module_extensions/cargo_bazel_bootstrap.bzl", line 66, column 17, in run
                fail("%s returned with exit code %d:\n%s" % (pretty_args, result.return_code, result.stderr))
Error in fail: /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/modextwd/rules_rust~~crate/cargo-bazel splice --output-dir /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/modextwd/rules_rust~~crate/crates/splicing-output --config /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/modextwd/rules_rust~~crate/crates/config.json --splicing-manifest /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/modextwd/rules_rust~~crate/crates/splicing_manifest.json --cargo-lockfile /data/mm64/zhihaopan/starpls/Cargo.lock --cargo /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/external/rules_rust~~rust_host_tools~rust_host_tools/bin/cargo --rustc /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/external/rules_rust~~rust_host_tools~rust_host_tools/bin/rustc returned with exit code 1:
/data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/modextwd/rules_rust~~crate/cargo-bazel: /usr/lib64/libc.so.6: version `GLIBC_2.18' not found (required by /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/modextwd/rules_rust~~crate/cargo-bazel)
ERROR: Analysis of target '//editors/code:copy_starpls' failed; build aborted: error evaluating module extension crate in @@rules_rust~//crate_universe:extension.bzl
INFO: Elapsed time: 7.298s, Critical Path: 0.04s
INFO: 1 process: 1 internal.
ERROR: Build did NOT complete successfully
ERROR: Build failed. Not running target
FAILED:
    Fetching module extension crate in @@rules_rust~//crate_universe:extension.bzl; starting
```

**Comment by withered-magic** [comment]
Hmm, I'd probably have to try and repro this on centos 7 myself, but in the meantime what happens if you try building through cargo? e.g. 

```
cargo build --release # This builds the executable at ./target/release/starpls
```

**Comment by withered-magic** [comment]
https://github.com/withered-magic/starpls/pull/335 should fix this, will merge and cut a release later today

**Comment by lasorda** [comment]
> #335 should fix this, will merge and cut a release later today

I tried this, but it't not work
```
~/starpls git:(main)
➜ ./bazel-7.4.1-linux-x86_64 run -c opt //editors/code:copy_starpls
WARNING: Build option --extra_toolchains has changed, discarding analysis cache (this can be expensive, see https://bazel.build/advanced/performance/iteration-speed).
ERROR: /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/external/rules_rust~/crate_universe/private/module_extensions/cargo_bazel_bootstrap.bzl:66:17: Traceback (most recent call last):
        File "/data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/external/rules_rust~/crate_universe/extension.bzl", line 369, column 37, in _crate_impl                                                                                                     _generate_hub_and_spokes(module_ctx, cargo_bazel, cfg, annotations, cargo_lockfile = cargo_lockfile, manifests = manifests)                                                                                                                                           File "/data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/external/rules_rust~/crate_universe/extension.bzl", line 132, column 16, in _generate_hub_and_spokes                                                                                        cargo_bazel(splice_args)                                                                                                                                                                                                                                              File "/data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/external/rules_rust~/crate_universe/private/module_extensions/cargo_bazel_bootstrap.bzl", line 66, column 17, in run                                                                        fail("%s returned with exit code %d:\n%s" % (pretty_args, result.return_code, result.stderr))
Error in fail: /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/modextwd/rules_rust~~crate/cargo-bazel splice --output-dir /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/modextwd/rules_rust~~crate/crates/splicing-output --config /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/modextwd/rules_rust~~crate/crates/config.json --splicing-manifest /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/modextwd/rules_rust~~crate/crates/splicing_manifest.json --cargo-lockfile /data/mm64/zhihaopan/starpls/Cargo.lock --cargo /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/external/rules_rust~~rust_host_tools~rust_host_tools/bin/cargo --rustc /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/external/rules_rust~~rust_host_tools~rust_host_tools/bin/rustc returned with exit code 1:
/data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/modextwd/rules_rust~~crate/cargo-bazel: /usr/lib64/libc.so.6: version `GLIBC_2.18' not found (required by /data/mm64/zhihaopan/.cache/bazel/_bazel_zhihaopan/1e18932b4feeb2df13538edfa1b6593d/modextwd/rules_rust~~crate/cargo-bazel)
ERROR: Analysis of target '//editors/code:copy_starpls' failed; build aborted: error evaluating module extension crate in @@rules_rust~//crate_universe:extension.bzl
INFO: Elapsed time: 0.191s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
ERROR: Build did NOT complete successfully
ERROR: Build failed. Not running target
~/starpls git:(main)
➜ gd .

Δ .bazelrc
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────                                                                                                                                                                                                                                                                              ────────────────────────────────┐
• 24: test --test_output=errors │
────────────────────────────────┘
│ 24 │build:linux --sandbox_add_mount_pair=/tmp                                                                                        │ 24 │build:linux --sandbox_add_mount_pair=/tmp
│ 25 │build:linux --incompatible_enable_cc_toolchain_resolution                                                                        │ 25 │build:linux --incompatible_enable_cc_toolchain_resolution
│ 26 │build:linux --action_env BAZEL_DO_NOT_DETECT_CPP_TOOLCHAIN=1                                                                     │ 26 │build:linux --action_env BAZEL_DO_NOT_DETECT_CPP_TOOLCHAIN=1
│ 27 │build:linux --extra_toolchains @zig_sdk//toolchain:linux_arm64_gnu.2.28,@zig_sdk//toolchain:linux_amd64_gnu.2.28                 │ 27 │build:linux --extra_toolchains @zig_sdk//toolchain:linux_arm64_gnu.2.17,@zig_sdk//toolchain:linux_amd64_gnu.2.17
│ 28 │                                                                                                                                 │ 28 │
│ 29 │build:windows --nolegacy_external_runfiles                                                                                       │ 29 │build:windows --nolegacy_external_runfiles
~/starpls git:(main)
➜
```

**Comment by withered-magic** [comment]
Hmm, I'd probably have to debug the Bazel build on a CentOS machine myself, but at first glance it looks like the prebuilt `cargo-bazel` tool used by `rules_rust` was linked against a version of glibc that's incompatible with your machine.

In the meantime, can you try the new binaries released as part of v0.1.18 [here](https://github.com/withered-magic/starpls/releases/tag/v0.1.18)? I tried them on the `centos:7` image and it seems to work fine.

**Comment by lasorda** [comment]
> Hmm, I'd probably have to debug the Bazel build on a CentOS machine myself, but at first glance it looks like the prebuilt `cargo-bazel` tool used by `rules_rust` was linked against a version of glibc that's incompatible with your machine.
> 
> In the meantime, can you try the new binaries released as part of v0.1.18 [here](https://github.com/withered-magic/starpls/releases/tag/v0.1.18)? I tried running them in the `centos:7` Docker image and it seems to work fine.

v0.1.18 works!

**Comment by withered-magic** [comment]
Cool, thanks for confirming! I'll close this issue for now and follow up on the centos build separately
